### PR TITLE
[ITensors] [ENHANCMENT] Introduce `removeqn` function for removing a specified quantum number

### PR DESCRIPTION
**Pull request by mtfishman**
# Description

Introduce `removeqn` function for removing a specified quantum number.

<details><summary>Minimal demonstration of new behavior</summary><p>

```julia
julia> using ITensors

julia> s = siteind("Electron"; conserve_qns=true);

julia> T = op("c†↑", s);

julia> @show T;
T = ITensor ord=2
Dim 1: (dim=4|id=685|"Electron,Site")' <Out>
 1: QN(("Nf",0,-1),("Sz",0)) => 1
 2: QN(("Nf",1,-1),("Sz",1)) => 1
 3: QN(("Nf",1,-1),("Sz",-1)) => 1
 4: QN(("Nf",2,-1),("Sz",0)) => 1
Dim 2: (dim=4|id=685|"Electron,Site") <In>
 1: QN(("Nf",0,-1),("Sz",0)) => 1
 2: QN(("Nf",1,-1),("Sz",1)) => 1
 3: QN(("Nf",1,-1),("Sz",-1)) => 1
 4: QN(("Nf",2,-1),("Sz",0)) => 1
NDTensors.BlockSparse{Float64, Vector{Float64}, 2}
 4×4
Block(2, 1)
 [2:2, 1:1]
 1.0

Block(4, 3)
 [4:4, 3:3]
 1.0

julia> R = removeqn(T, "Sz");

julia> @show R;
R = ITensor ord=2
Dim 1: (dim=4|id=685|"Electron,Site")' <Out>
 1: QN("Nf",0,-1) => 1
 2: QN("Nf",1,-1) => 2
 3: QN("Nf",2,-1) => 1
Dim 2: (dim=4|id=685|"Electron,Site") <In>
 1: QN("Nf",0,-1) => 1
 2: QN("Nf",1,-1) => 2
 3: QN("Nf",2,-1) => 1
NDTensors.BlockSparse{Float64, Vector{Float64}, 2}
 4×4
Block(2, 1)
 [2:3, 1:1]
 1.0
 0.0

Block(3, 2)
 [4:4, 2:3]
 0.0  1.0
```
</p></details>

# How Has This Been Tested?

Tests were added checking the elements are the same before and after and that the QNs are the ones we expect.

# Checklist:

- [x] My code follows the style guidelines of this project. Please run `using JuliaFormatter; format(".")` in the base directory of the repository (`~/.julia/dev/ITensors`) to format your code according to our style guidelines.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that verify the behavior of the changes I made.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

**Comment by emstoudenmire** [comment]
Nice to have this! I can see how it can be very useful.

Looking at the code, I guess right now this doesn't merge blocks that end up having the same QN after this procedure correct? That's not necessarily a problem, as we saw with the "splitblocks" experiments with DMRG where packing this as much as possible isn't always an advantage, but I just wanted to understand if that's done here.

**Comment by mtfishman** [comment]
It merges contiguous blocks (you can see that in the example in the first post). There is an option to turn that off with the keyword argument `mergeblocks=false`.

It doesn't permute the blocks, so sometimes it won't fully merge the blocks, for example if you remove the `"Nf"` QN in the example in the first post it won't merge the `("Sz", 0)` sectors since they are not next to each other (they are the first and last states for the `"Electron"` type, as you know). The reason for that is I didn't want this function to produce indices with new Index ids (so that you could for example remove a QN from both an MPS and MPO and keep the indices compatible). It seems ok to keep the Index id the same if QN are just merged as long as the states remain in the original ordering (i.e. the tensor elements are the same, just the sparsity pattern changes).

**Comment by emstoudenmire** [comment]
Yes that's the right behavior that we want I agree.

What I was asking about also was at the storage level. Like are the number of separate blocks in the storage changed at all? Or it's just kept the same and everything still works as long as the new index values are in 1-1 correspondence with the old ones, even if a single QN block now spans over two different blocks in the storage?

**Comment by mtfishman** [comment]
With `mergeblocks=false`, the block structure remains the same (same number of blocks of the same size, just with different QNs).

With `mergeblocks=true` (the default), if there are contiguous blocks with the same QNs after QNs are removed, blocks are merged, so it results in a tensor with fewer blocks and for blocks that were merged they are larger and filled out with zeros.

**Comment by mtfishman** [comment]
Let me know if that sounds reasonable to you, if so I'll merge and register a new version.

**Comment by emstoudenmire** [comment]
Great. Yes please go ahead and merge this PR. 

I was wondering, just for my understanding, which line of code modifies the tensor storage, since I could only see code here that was modifying indices.

**Comment by mtfishman** [comment]
The current implementation is pretty simplistic. It make a new ITensor with empty storage using the new indices and then sets the elements one by one. In this way it dynamically allocates blocks that are nonzero as needed. This is more of a "quick and dirty" solution which could be optimized (for example a simple optimization in the `mergeblocks=false` case would be to just copy the storage). However I doubt this would be a bottleneck in a calculation, I imagine it is usually just done once before doing some measurements or a new DMRG run.

File: src/qn/qnitensor.jl [hunk 1]
```diff
@@ -481,3 +484,17 @@ function splitblocks(A::ITensor, is=inds(A); tol=0)
...
+  if !hasqns(T)
+    return T
+  end
+  inds_R = removeqn(inds(T), qn_name; mergeblocks)
+  R = ITensor(inds_R)
+  for iv in eachindex(T)
+    if !iszero(T[iv])
+      R[iv] = T[iv]
+    end
+  end
```
Comment (mtfishman) [inline src/qn/qnitensor.jl:498]: This is where the new tensor with the new storage is being made.

Comment (emstoudenmire) [reply src/qn/qnitensor.jl:498]: Ok thanks! I kept missing it when looking through the changes. Yes later on I imagine we can have better facilities for slicing and then could do sliced assignments of entire blocks of T into either blocks or sub-blocks of R. (As a side thought, better slicing for both dense and block sparse tensors might simplify the block sparse combiner code.)

Comment (mtfishman) [reply src/qn/qnitensor.jl:498]: Yeah, this would be a good candidate for making use of better slicing operations.

**Pull request merged**